zcash_client_sqlite: delete_account fails if cross-account transactions exist

Originally reported by @noop-sk:
```
Bug Report: delete_account fails with Invalid parameter name: :address when sent_notes reference the deleted account

  Affects: zcash_client_sqlite 0.20.2 and 0.21.0
  Severity: High — renders Keystone hardware wallet disconnect permanently broken for affected users
  Reproducible: State-dependent (see trigger condition below)

  ---
  Summary

  WalletDb::delete_account panics with rusqlite::Error::InvalidParameterName(":address") when the account being deleted is referenced by any sent_notes row. The SQL statement uses the parameter name
  :to_address, but the binding code passes :address. This mismatch is silent when the loop body never executes (no matching rows), which is why the bug went unnoticed — it only manifests once a user has
  made a cross-account internal transfer to the Keystone account.

  ---
  Location

  zcash_client_sqlite/src/wallet.rs, function delete_account:

  let mut update_sent_note = conn.prepare(
      r#"
      UPDATE sent_notes
      SET to_address = :to_address, to_account_id = NULL   // ← expects :to_address
      WHERE id = :sent_note_id
      "#,
  )?;

  // ...

  while let Some(row) = rows.next()? {
      if let Some(address) = row.get::<_, Option<String>>("to_address")? {
          update_sent_note.execute(named_params![
              ":sent_note_id": row.get::<_, i64>("sent_note_id")?,
              ":address": address   // ← BUG: should be ":to_address"
          ])?;
      }
  }

  ---
  Fix

  One-line change:

  // Before
  ":address": address

  // After
  ":to_address": address

  ---
  Trigger Condition

  The bug is only reachable when the following query returns at least one row:

  SELECT sn.id AS sent_note_id, ...
  FROM sent_notes sn
  JOIN v_received_outputs ro ON ro.sent_note_id = sn.id
  JOIN addresses ON addresses.id = ro.address_id
  JOIN accounts ta ON ta.id = [sn.to](http://sn.to/)_account_id
  WHERE ta.uuid = :account_uuid   -- the account being deleted

  In practice this means: a transaction was sent to an address belonging to the account being deleted, and that output was received and scanned. In a multi-account wallet (e.g. Zashi account + Keystone
  hardware wallet), this state is reached when the user sends ZEC from account A to an address of account B, or via auto-shielding of transparent Keystone funds.

  ---
  Observed Crash (Android)

  java.lang.RuntimeException: Invalid parameter name: :address
      at cash.z.ecc.android.sdk.internal.jni.RustBackend.deleteAccount(Native Method)
      ...
  Caused by: java.lang.RuntimeException: Invalid parameter name: :address
      at cash.z.ecc.android.sdk.internal.jni.RustBackend.deleteAccount(Native Method)

  Full JNI call chain:
  Synchronizer.deleteAccount(AccountUuid) → TypesafeBackendImpl.deleteAccount → RustBackend.deleteAccount (JNI) → wallet::delete_account → rusqlite InvalidParameterName(":address")

  App version: 3.6.0 (SDK 2.6.2-SNAPSHOT, zcash_client_sqlite 0.21.0, mainnet)

  ---
  Why It Appeared "Suddenly"

  The defect has been present since at least zcash_client_sqlite 0.20.2. delete_account succeeded for all users whose Keystone account had never appeared as to_account_id in sent_notes. The first
  cross-account transfer (internal send, auto-shield of Keystone transparent balance, or any multi-account operation) creates the triggering database state, after which every subsequent disconnect attempt
  fails permanently.

  ---
  Impact

  Users cannot disconnect their Keystone hardware wallet. The operation fails at the native layer and the error surfaces to the UI. The wallet database is left in a consistent state (the account is not
  deleted), so there is no data loss — but the account cannot be removed.
```

--- a/zcash_client_backend/src/data_api/testing/pool.rs
+++ b/zcash_client_backend/src/data_api/testing/pool.rs
@@ -3295,6 +3295,110 @@ where
     );
 }
 
+/// Regression test for a bug in which [`WalletWrite::delete_account`] failed with a
+/// `rusqlite::Error::InvalidParameterName(":address")` panic when the account being
+/// deleted was referenced by a `sent_notes` row via its `to_account_id` column.
+///
+/// The triggering state is reached when a transaction is sent from one account in the
+/// wallet to an address belonging to a second account in the same wallet, and the
+/// transaction is then decrypted via [`decrypt_and_store_transaction`] so that the
+/// cross-account transfer is recorded with a non-null `to_account_id` and a received
+/// output that has an associated address. Deleting the recipient account then exercises
+/// the `sent_notes` update path inside `delete_account`.
+///
+/// [`WalletWrite::delete_account`]: crate::data_api::WalletWrite::delete_account
+pub fn account_deletion_with_internal_transfer<T: ShieldedPoolTester, DSF>(
+    ds_factory: DSF,
+    cache: impl TestCache,
+) where
+    DSF: DataStoreFactory,
+    <DSF as DataStoreFactory>::DataStore: Reset,
+{
+    let mut st = TestBuilder::new()
+        .with_data_store_factory(ds_factory)
+        .with_block_cache(cache)
+        .build();
+
+    // Add two accounts to the wallet, derived from the same seed.
+    let seed = Secret::new([0u8; 32].to_vec());
+    let birthday = AccountBirthday::from_sapling_activation(st.network(), BlockHash([0; 32]));
+    let (account1, usk1) = st
+        .wallet_mut()
+        .create_account("account1", &seed, &birthday, None)
+        .unwrap();
+    let dfvk1 = T::sk_to_fvk(T::usk_to_sk(&usk1));
+
+    let (account2, usk2) = st
+        .wallet_mut()
+        .create_account("account2", &seed, &birthday, None)
+        .unwrap();
+    let dfvk2 = T::sk_to_fvk(T::usk_to_sk(&usk2));
+
+    // Add funds to account 1 in a single note.
+    let value = Zatoshis::from_u64(100000).unwrap();
+    let (h, _, _) = st.generate_next_block(&dfvk1, AddressType::DefaultExternal, value);
+    st.scan_cached_blocks(h, 1);
+
+    assert_eq!(st.get_total_balance(account1), value);
+    assert_eq!(st.get_total_balance(account2), Zatoshis::ZERO);
+
+    // Send funds from account 1 to an address belonging to account 2.
+    let bal_2 = Zatoshis::from_u64(50000).unwrap();
+    let addr2 = T::fvk_default_address(&dfvk2);
+    let req = TransactionRequest::new(vec![Payment::without_memo(
+        addr2.to_zcash_address(st.network()),
+        bal_2,
+    )])
+    .unwrap();
+
+    let change_strategy = fees::standard::SingleOutputChangeStrategy::new(
+        StandardFeeRule::Zip317,
+        None,
+        T::SHIELDED_PROTOCOL,
+        DustOutputPolicy::default(),
+    );
+    let input_selector = GreedyInputSelector::new();
+
+    let txid = st
+        .spend(
+            &input_selector,
+            &change_strategy,
+            &usk1,
+            req,
+            OvkPolicy::Sender,
+            ConfirmationsPolicy::MIN,
+        )
+        .unwrap()[0];
+
+    let (h, _) = st.generate_next_block_including(txid);
+    st.scan_cached_blocks(h, 1);
+
+    assert_eq!(st.get_total_balance(account2), bal_2);
+
+    // Decrypt and store the transaction. Because the wallet owns the funding inputs
+    // (account 1) and the output is received by account 2, this records the send as an
+    // internal cross-account transfer, setting `sent_notes.to_account_id` to account 2
+    // and associating the received output with account 2's address. This is the state
+    // that triggers the `delete_account` bug.
+    let tx = st.wallet().get_transaction(txid).unwrap().unwrap();
+    let params = *st.network();
+    decrypt_and_store_transaction(&params, st.wallet_mut(), &tx, Some(h)).unwrap();
+
+    // Deleting account 2, the recipient of the internal transfer, must succeed. Prior to
+    // the fix this failed with `rusqlite::Error::InvalidParameterName(":address")` because
+    // the `sent_notes` update statement bound the wrong parameter name.
+    assert_matches!(st.wallet_mut().delete_account(account2), Ok(_));
+
+    // account 1 should still exist and retain its change balance.
+    let summary = st
+        .wallet()
+        .get_wallet_summary(ConfirmationsPolicy::MIN)
+        .unwrap()
+        .unwrap();
+    assert!(summary.account_balances().get(&account2).is_none());
+    assert!(summary.account_balances().contains_key(&account1));
+}
+
 pub fn external_address_change_spends_detected_in_restore_from_seed<T: ShieldedPoolTester, Dsf>(
     ds_factory: Dsf,
     cache: impl TestCache,

--- a/zcash_client_sqlite/CHANGELOG.md
+++ b/zcash_client_sqlite/CHANGELOG.md
@@ -8,6 +8,16 @@ indicated by the `PLANNED` status in order to make it possible to correctly
 represent the transitive `semver` implications of changes within the enclosing
 workspace.
 
+## Unreleased
+
+### Fixed
+- Fixed a bug in `WalletDb::delete_account` that caused it to fail with
+  `rusqlite::Error::InvalidParameterName(":address")` when the account being
+  deleted was referenced by a `sent_notes` row via its `to_account_id` column
+  (for example, after an internal transfer to an address belonging to the
+  account being deleted). The `sent_notes` update statement bound a parameter
+  named `:address` while the SQL expected `:to_address`.
+
 ## [0.21.0] - 2026-06-02
 
 ### Changed

--- a/zcash_client_sqlite/src/testing/pool.rs
+++ b/zcash_client_sqlite/src/testing/pool.rs
@@ -205,6 +205,13 @@ pub(crate) fn account_deletion<T: ShieldedPoolTester>() {
     )
 }
 
+pub(crate) fn account_deletion_with_internal_transfer<T: ShieldedPoolTester>() {
+    zcash_client_backend::data_api::testing::pool::account_deletion_with_internal_transfer::<T, _>(
+        TestDbFactory::default(),
+        BlockCache::new(),
+    )
+}
+
 pub(crate) fn external_address_change_spends_detected_in_restore_from_seed<
     T: ShieldedPoolTester,
 >() {

--- a/zcash_client_sqlite/src/wallet.rs
+++ b/zcash_client_sqlite/src/wallet.rs
@@ -713,7 +713,7 @@ pub(crate) fn delete_account(
         if let Some(address) = row.get::<_, Option<String>>("to_address")? {
             update_sent_note.execute(named_params![
                 ":sent_note_id": row.get::<_, i64>("sent_note_id")?,
-                ":address": address
+                ":to_address": address
             ])?;
         }
     }

--- a/zcash_client_sqlite/src/wallet/orchard.rs
+++ b/zcash_client_sqlite/src/wallet/orchard.rs
@@ -607,6 +607,11 @@ pub(crate) mod tests {
     }
 
     #[test]
+    fn account_deletion_with_internal_transfer() {
+        testing::pool::account_deletion_with_internal_transfer::<OrchardPoolTester>()
+    }
+
+    #[test]
     fn external_address_change_spends_detected_in_restore_from_seed() {
         testing::pool::external_address_change_spends_detected_in_restore_from_seed::<
             OrchardPoolTester,

--- a/zcash_client_sqlite/src/wallet/sapling.rs
+++ b/zcash_client_sqlite/src/wallet/sapling.rs
@@ -550,6 +550,11 @@ pub(crate) mod tests {
     }
 
     #[test]
+    fn account_deletion_with_internal_transfer() {
+        testing::pool::account_deletion_with_internal_transfer::<SaplingPoolTester>()
+    }
+
+    #[test]
     fn external_address_change_spends_detected_in_restore_from_seed() {
         testing::pool::external_address_change_spends_detected_in_restore_from_seed::<
             SaplingPoolTester,

